File: src/clojure/advent_of_code_2015/day_20.clj
(ns advent-of-code-2015.day-20
  (:import [advent_of_code_2015 Day20])
  (:gen-class))

(defn sigma
  [n]
  (reduce + (filter #(zero? (mod n %)) (range 1 (inc n)))))

(defn sigma-lower-bound
  "Given a value of sigma_1(n) computes a lower bound for the value of n."
  [sigma]
  (long (Math/floor (Math/sqrt sigma))))

(defn sigma-upper-bound
  "Given a value of sigma_1(n) computes an upper bound for the value of n."
  [sigma]
  (long (Math/ceil
         (+ (* 0.5 (+ 1.0 (* 2.0 sigma)))
            (* -0.5 (Math/sqrt (+ 1.0 (* 4.0 sigma))))))))

(defn part-1
  [bound]
  (let [sigma (quot bound 10)]
    (first
     (drop-while #(< (Day20/sigma %) sigma)
                 (range (sigma-lower-bound sigma) (inc (sigma-upper-bound sigma)))))))

(defn part-2
  [bound]
  (let [sigma (quot bound 11)]
    (first (drop-while #(< (Day20/sigma_2 %) sigma)
                       (range (sigma-lower-bound sigma)
                              (sigma-upper-bound sigma))))))

(defn day-20
  [input]
  (println "Day 20")
  (println "House number: " (part-1 (Integer. input)))
  (println "New house number: " (part-2 (Integer. input))))

File: src/java/advent_of_code_2015/Day20.java
package advent_of_code_2015;

public class Day20 {
    /// Calculates the sum of the divisors of n.
    public static int sigma(int n) {
        int total = 1;

        for(int d = 2; d <= n; d++) {
            if (n % d == 0) {
                total += d;
            }
        }

        return total;
    }

    public static int sigma_2(int n) {
        int total = 0;

        for(int d = n; d > 0; d--) {
            if (n / d > 50) {
                break;
            }
            if (n % d == 0) {
                total += d;
            }
        }

        return total;
    }
}
